wren-ui/src/apollo/server/services/dashboardService.ts

```diff
@@ -193,4 +254,270 @@ export class DashboardService implements IDashboardService {
+        day: DAYS[adjustedDayIndex],
+        timezone,
+      };
+    }
+
+    // Return a new schedule object with local hours and minutes
+    return {
+      ...schedule,
+      hour: localHour,
+      minute: localMinute,
+      timezone,
+    };
+  }
+
+  protected generateCronExpression(schedule: DashboardSchedule): string | null {
+    const { frequency, day, hour, minute } = this.toUTC(schedule);
+
+    switch (frequency) {
+      case ScheduleFrequencyEnum.DAILY:
+        return `${minute} ${hour} * * *`;
+      case ScheduleFrequencyEnum.WEEKLY:
+        // e.g. 0 10 * * MON
+        return `${minute} ${hour} * * ${day}`;
+      case ScheduleFrequencyEnum.CUSTOM:
+        return schedule.cron;
+      case ScheduleFrequencyEnum.NEVER:
+        return null;
+      default:
+        logger.warn(`Unsupported schedule frequency: ${frequency}`);
+        return null;
+    }
+  }
+
+  protected calculateNextRunTime(cronExpression: string): Date | null {
+    try {
+      const interval = CronExpressionParser.parse(cronExpression, {
+        currentDate: new Date(),
+      });
+      return interval.next().toDate();
+    } catch (error) {
+      logger.error(`Failed to parse cron expression: ${error.message}`);
+      return null;
+    }
+  }
+
+  protected validateScheduleInput(data: SetDashboardCacheData): void {
+    const { schedule } = data;
+
+    if (schedule.frequency === ScheduleFrequencyEnum.WEEKLY && !schedule.day) {
+      throw new Error('Day of week is required for weekly schedule');
+    }
+
+    if (schedule.frequency === ScheduleFrequencyEnum.CUSTOM && !schedule.cron) {
+      throw new Error('Cron expression is required for custom schedule');
+    }
+
+    if (schedule.hour < 0 || schedule.hour > 23) {
+      throw new Error('Hour must be between 0 and 23');
+    }
+
+    if (schedule.minute < 0 || schedule.minute > 59) {
+      throw new Error('Minute must be between 0 and 59');
+    }
+
+    if (schedule.timezone) {
+      try {
+        new Date().toLocaleString('en-US', { timeZone: schedule.timezone });
+      } catch (_) {
+        throw new Error(`Invalid timezone: ${schedule.timezone}`);
+      }
+    }
+
+    if (schedule.frequency === ScheduleFrequencyEnum.CUSTOM) {
+      // can not less than 10 minutes, skip if is local
+      if (process.env.NODE_ENV === 'development') return;
+      const baseInterval = CronExpressionParser.parse(schedule.cron, {
+        currentDate: new Date(),
+      });
+      const baseDate = baseInterval.next().toDate();
+
+      const nextInterval = CronExpressionParser.parse(schedule.cron, {
+        currentDate: baseDate,
+      });
+      const nextDate = nextInterval.next().toDate();
+      const diff = nextDate.getTime() - baseDate.getTime();
+      if (diff < 10 * 60 * 1000) {
+        throw new Error(
+          'Custom cron expression must be at least 10 minutes apart',
+        );
+      }
+    }
+  }
+
+  public parseCronExpression(dashboard: Dashboard): DashboardSchedule {
```

Comment:
Feels like `parseCronExpression` and `toTimezone` (maybe other methods too) could be put into a utility file, instead of here in service level.